[quant] Add QConfig v2 keys

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94251

Summary:
This PR adds keys for the second version of QConfig that allows users have more fine grained quantization
settings. version 1 and version 2 keys can't be set at the same time.

QConfig is an object for users to set the observer or fake quantize classes for input/output/weight of an operator/module.

**QConfig v1**
QConfig v1 (current QConfig) was introduced when eager mode quantization is designed,
it has two keys: "activation" and "weight", here
activation means the output activation, and weight means weight attributes for weighted modules

Later we developed torchscript graph mode quantization (quantize_jit) and fx graph
mode quantization (quantize_fx), and now quantization in pytorch 2.0 export path
(quantize_pt2e), but we did not redesign QConfig, and "activation" is redefined as a
setting for both input and output, "weight" also expanded it's definition serving both setting for module\
weigt and the weight input of torch functional or torch ops (with the aid of extra
configurations to specify which input is weight) and we do not have a way to configure
the observer/fake_quant constructor for bias.

**QConfig v2**
This PR adds keys for the second version of QConfig, it has the following keys:
["input_args", "module_attrs", "output", "version"]

"input_args": is used to configure the observer/fakequant constructors for input arguments,
it should have the same structure as the positional arguments e.g. for an operator with
the following signature: `op(arg0, arg1, arg2)`, "input_args" should look like the following:
```
qconfig.input_args = (obs0, obs1, obs2)
```
this can be expanded to more general structures like nested dictionary with a tuple as a value etc.

"module_attrs": is used to configure the observer/fakequant constructors for module attributes,
 e.g. weight, it should be a dictionary from module attribute name to the observer/fake_quant class, e.g.:

qconfig.module_attrs = {"weight": obs0}

"output": is used to configure the observer/fakequant constructors for output,
it should match the structure of output, e.g. if output has the following structure:
{"key0": out0, "key1": (out1, out2)}, we will have the following output config:
qconfig.output = {"key0": obs0, "key1": (obs1, obs2)}

"version": should be None (meaning v1), 1 or 2

**Examples**::

  # v1
  my_qconfig = QConfig(
      activation=MinMaxObserver.with_args(dtype=torch.qint8),
      weight=default_observer.with_args(dtype=torch.qint8))
  # v2
  my_qconfig = QConfig(
      input_args=(MinMaxObserver.with_args(dtype=torch.qint8),),
      module_attrs={"weight": default_observer.with_args(dtype=torch.qint8)},
      output=(MinMaxObserver.with_args(dtype=torch.qint8),)
  )

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
[quant] Add QConfig v2 keys

Summary:
This PR adds keys for the second version of QConfig that allows users have more fine grained quantization
settings. version 1 and version 2 keys can't be set at the same time.

QConfig is an object for users to set the observer or fake quantize classes for input/output/weight of an operator/module.

**QConfig v1**
QConfig v1 (current QConfig) was introduced when eager mode quantization is designed,
it has two keys: "activation" and "weight", here
activation means the output activation, and weight means weight attributes for weighted modules

Later we developed torchscript graph mode quantization (quantize_jit) and fx graph
mode quantization (quantize_fx), and now quantization in pytorch 2.0 export path
(quantize_pt2e), but we did not redesign QConfig, and "activation" is redefined as a
setting for both input and output, "weight" also expanded it's definition serving both setting for module\
weigt and the weight input of torch functional or torch ops (with the aid of extra
configurations to specify which input is weight) and we do not have a way to configure
the observer/fake_quant constructor for bias.

**QConfig v2**
This PR adds keys for the second version of QConfig, it has the following keys:
["input_args", "module_attrs", "output", "version"]

"input_args": is used to configure the observer/fakequant constructors for input arguments,
it should have the same structure as the positional arguments e.g. for an operator with
the following signature: `op(arg0, arg1, arg2)`, "input_args" should look like the following:
```
qconfig.input_args = (obs0, obs1, obs3)
```
this can be expanded to more general structures like nested dictionary with a tuple as a value etc.

"module_attrs": is used to configure the observer/fakequant constructors for module attributes,
 e.g. weight, it should be a dictionary from module attribute name to the observer/fake_quant class, e.g.:

qconfig.module_attrs = {"weight": obs0}

"output": is used to configure the observer/fakequant constructors for output,
it should match the structure of output, e.g. if output has the following structure:
{"key0": out0, "key1": (out1, out2)}, we will have the following output config:
qconfig.output = {"key0": obs0, "key1": (obs1, obs2)}

"version": should be None (meaning v1), 1 or 2

**Examples**::

  # v1
  my_qconfig = QConfig(
      activation=MinMaxObserver.with_args(dtype=torch.qint8),
      weight=default_observer.with_args(dtype=torch.qint8))
  # v2
  my_qconfig = QConfig(
      input_args=(MinMaxObserver.with_args(dtype=torch.qint8),),
      module_attrs={"weight": default_observer.with_args(dtype=torch.qint8)},
      output=(MinMaxObserver.with_args(dtype=torch.qint8),)
  )

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

leslie-fang-intel commented on torch/ao/quantization/qconfig.py:
````diff
@@ -94,13 +97,68 @@ class QConfig(namedtuple('QConfig', ['activation', 'weight'])):
+    it should have the same structure as the positional arguments e.g. for an operator with
+    the following signature: `op(arg0, arg1, arg2)`, "input_args" should look like the following:
+    ```
+    qconfig.input_args = (obs0, obs1, obs3)
````
is it a typo error? should it be `qconfig.input_args = (obs0, obs1, obs2)`?

jerryzh168 replied:
yeah, will change

leslie-fang-intel commented on torch/ao/quantization/qconfig.py:
```diff
@@ -94,13 +97,68 @@ class QConfig(namedtuple('QConfig', ['activation', 'weight'])):
+    This PR adds keys for the second version of QConfig, it has the following keys:
+    ["input_args", "module_attrs", "output", "version"]
+
+    "input_args": is used to configure the observer/fakequant constructors for input arguments,
```
So for the `QConfig v2`, the weight input of torch functional will be set here instead of the `module_attrs`.

jerryzh168 replied:
it should be a part of `input_args`, we can add `input_kwargs` later if needed

jerryzh168 committed:
Update on "[quant] Add QConfig v2 keys"


Summary:
This PR adds keys for the second version of QConfig that allows users have more fine grained quantization
settings. version 1 and version 2 keys can't be set at the same time.

QConfig is an object for users to set the observer or fake quantize classes for input/output/weight of an operator/module.

**QConfig v1**
QConfig v1 (current QConfig) was introduced when eager mode quantization is designed,
it has two keys: "activation" and "weight", here
activation means the output activation, and weight means weight attributes for weighted modules

Later we developed torchscript graph mode quantization (quantize_jit) and fx graph
mode quantization (quantize_fx), and now quantization in pytorch 2.0 export path
(quantize_pt2e), but we did not redesign QConfig, and "activation" is redefined as a
setting for both input and output, "weight" also expanded it's definition serving both setting for module\
weigt and the weight input of torch functional or torch ops (with the aid of extra
configurations to specify which input is weight) and we do not have a way to configure
the observer/fake_quant constructor for bias.

**QConfig v2**
This PR adds keys for the second version of QConfig, it has the following keys:
["input_args", "module_attrs", "output", "version"]

"input_args": is used to configure the observer/fakequant constructors for input arguments,
it should have the same structure as the positional arguments e.g. for an operator with
the following signature: `op(arg0, arg1, arg2)`, "input_args" should look like the following:
```
qconfig.input_args = (obs0, obs1, obs2)
```
this can be expanded to more general structures like nested dictionary with a tuple as a value etc.

"module_attrs": is used to configure the observer/fakequant constructors for module attributes,
 e.g. weight, it should be a dictionary from module attribute name to the observer/fake_quant class, e.g.:

qconfig.module_attrs = {"weight": obs0}

"output": is used to configure the observer/fakequant constructors for output,
it should match the structure of output, e.g. if output has the following structure:
{"key0": out0, "key1": (out1, out2)}, we will have the following output config:
qconfig.output = {"key0": obs0, "key1": (obs1, obs2)}

"version": should be None (meaning v1), 1 or 2

**Examples**::

  # v1
  my_qconfig = QConfig(
      activation=MinMaxObserver.with_args(dtype=torch.qint8),
      weight=default_observer.with_args(dtype=torch.qint8))
  # v2
  my_qconfig = QConfig(
      input_args=(MinMaxObserver.with_args(dtype=torch.qint8),),
      module_attrs={"weight": default_observer.with_args(dtype=torch.qint8)},
      output=(MinMaxObserver.with_args(dtype=torch.qint8),)
  )

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented:
Overall this makes sense.

Thoughts on making it a separate object?  Since we have "users have to specify either v1 keys or v2 keys", the change from that to "QConfigV1 object and QConfigV2 object" is small.  Then, if needed, we could convert objects of type `QConfigV1` to objects of type `QConfigV2` under the hood and keep most internal code interacting with the latter version.

jerryzh168 committed:
Update on "[quant] Add QConfig v2 keys"


Summary:
This PR adds keys for the second version of QConfig that allows users have more fine grained quantization
settings. version 1 and version 2 keys can't be set at the same time.

QConfig is an object for users to set the observer or fake quantize classes for input/output/weight of an operator/module.

**QConfig v1**
QConfig v1 (current QConfig) was introduced when eager mode quantization is designed,
it has two keys: "activation" and "weight", here
activation means the output activation, and weight means weight attributes for weighted modules

Later we developed torchscript graph mode quantization (quantize_jit) and fx graph
mode quantization (quantize_fx), and now quantization in pytorch 2.0 export path
(quantize_pt2e), but we did not redesign QConfig, and "activation" is redefined as a
setting for both input and output, "weight" also expanded it's definition serving both setting for module\
weigt and the weight input of torch functional or torch ops (with the aid of extra
configurations to specify which input is weight) and we do not have a way to configure
the observer/fake_quant constructor for bias.

**QConfig v2**
This PR adds keys for the second version of QConfig, it has the following keys:
["input_args", "module_attrs", "output", "version"]

"input_args": is used to configure the observer/fakequant constructors for input arguments,
it should have the same structure as the positional arguments e.g. for an operator with
the following signature: `op(arg0, arg1, arg2)`, "input_args" should look like the following:
```
qconfig.input_args = (obs0, obs1, obs2)
```
this can be expanded to more general structures like nested dictionary with a tuple as a value etc.

"module_attrs": is used to configure the observer/fakequant constructors for module attributes,
 e.g. weight, it should be a dictionary from module attribute name to the observer/fake_quant class, e.g.:

qconfig.module_attrs = {"weight": obs0}

"output": is used to configure the observer/fakequant constructors for output,
it should match the structure of output, e.g. if output has the following structure:
{"key0": out0, "key1": (out1, out2)}, we will have the following output config:
qconfig.output = {"key0": obs0, "key1": (obs1, obs2)}

"version": should be None (meaning v1), 1 or 2

**Examples**::

  # v1
  my_qconfig = QConfig(
      activation=MinMaxObserver.with_args(dtype=torch.qint8),
      weight=default_observer.with_args(dtype=torch.qint8))
  # v2
  my_qconfig = QConfig(
      input_args=(MinMaxObserver.with_args(dtype=torch.qint8),),
      module_attrs={"weight": default_observer.with_args(dtype=torch.qint8)},
      output=(MinMaxObserver.with_args(dtype=torch.qint8),)
  )

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
> Overall this makes sense.
> 
> Thoughts on making it a separate object? Since we have "users have to specify either v1 keys or v2 keys", the change from that to "QConfigV1 object and QConfigV2 object" is small. Then, if needed, we could convert objects of type `QConfigV1` to objects of type `QConfigV2` under the hood and keep most internal code interacting with the latter version.

I don't have a strong preference, what would be the name for the second one? QConfigV2?

jerryzh168 committed:
Update on "[quant] Add QConfig v2 keys"


Summary:
This PR adds keys for the second version of QConfig that allows users have more fine grained quantization
settings. version 1 and version 2 keys can't be set at the same time.

QConfig is an object for users to set the observer or fake quantize classes for input/output/weight of an operator/module.

**QConfig v1**
QConfig v1 (current QConfig) was introduced when eager mode quantization is designed,
it has two keys: "activation" and "weight", here
activation means the output activation, and weight means weight attributes for weighted modules

Later we developed torchscript graph mode quantization (quantize_jit) and fx graph
mode quantization (quantize_fx), and now quantization in pytorch 2.0 export path
(quantize_pt2e), but we did not redesign QConfig, and "activation" is redefined as a
setting for both input and output, "weight" also expanded it's definition serving both setting for module\
weigt and the weight input of torch functional or torch ops (with the aid of extra
configurations to specify which input is weight) and we do not have a way to configure
the observer/fake_quant constructor for bias.

**QConfig v2**
This PR adds keys for the second version of QConfig, it has the following keys:
["input_args", "module_attrs", "output", "version"]

"input_args": is used to configure the observer/fakequant constructors for input arguments,
it should have the same structure as the positional arguments e.g. for an operator with
the following signature: `op(arg0, arg1, arg2)`, "input_args" should look like the following:
```
qconfig.input_args = (obs0, obs1, obs2)
```
this can be expanded to more general structures like nested dictionary with a tuple as a value etc.

"module_attrs": is used to configure the observer/fakequant constructors for module attributes,
 e.g. weight, it should be a dictionary from module attribute name to the observer/fake_quant class, e.g.:

qconfig.module_attrs = {"weight": obs0}

"output": is used to configure the observer/fakequant constructors for output,
it should match the structure of output, e.g. if output has the following structure:
{"key0": out0, "key1": (out1, out2)}, we will have the following output config:
qconfig.output = {"key0": obs0, "key1": (obs1, obs2)}

"version": should be None (meaning v1), 1 or 2

**Examples**::

  # v1
  my_qconfig = QConfig(
      activation=MinMaxObserver.with_args(dtype=torch.qint8),
      weight=default_observer.with_args(dtype=torch.qint8))
  # v2
  my_qconfig = QConfig(
      input_args=(MinMaxObserver.with_args(dtype=torch.qint8),),
      module_attrs={"weight": default_observer.with_args(dtype=torch.qint8)},
      output=(MinMaxObserver.with_args(dtype=torch.qint8),)
  )

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
After thinking about this a bit more, I think we don't need this, since it's not going to be general enough to cover all the cases anyway, we'll have the most general API in the quantization in pytorch 2.0 export (quantize_pt2e) api, that allows users to configure the observer/fq constructor for each node directly. for details about the reasoning of abandoning this approach see: https://docs.google.com/document/d/1CuBOcLnqOJ22urA3QYc8Kep2mdijv4DVr1v6tlk6Nfw/edit
